run/run.go

```diff
@@ -0,0 +1,70 @@
+package run
+
+import (
+	"bufio"
+	"io"
+)
+
+// Runner just runs anything.
+type Runner interface {
+	// StdoutPipe returns a pipe that will be connected to the runner's
+	// standard output when the command starts.
+	StdoutPipe() (io.ReadCloser, error)
+
+	// StderrPipe returns a pipe that will be connected to the runner's
+	// standard error when the command starts.
+	StderrPipe() (io.ReadCloser, error)
+
+	// Start starts the runner but does not wait for it to complete.
+	Start() error
+
+	// Wait waits for the runner to exit. It must have been started by Start.
+	//
+	// The returned error is nil if the runner has no problems copying
+	// stdin, stdout, and stderr, and exits with a zero exit status.
+	Wait() error
+}
+
+// NewShRunner creates a system standard shell script runner.
+func NewShRunner(code string, workdir string, env []string) (Runner, error) {
+	return newCmdRunner("sh", code, workdir, env)
+}
+
+// NewBashRunner creates a Bash runner.
+func NewBashRunner(code string, workdir string, env []string) (Runner, error) {
+	return newCmdRunner("bash", code, workdir, env)
+}
+
+// NewPythonRunner creates a Python runner.
+func NewPythonRunner(code string, workdir string, env []string) (Runner, error) {
+	return newCmdRunner("python", code, workdir, env)
+}
+
+// NewRubyRunner creates a Ruby runner.
+func NewRubyRunner(code string, workdir string, env []string) (Runner, error) {
+	return newCmdRunner("ruby", code, workdir, env)
+}
+
+// NewPerlRunner creates a Perl runner.
+func NewPerlRunner(code string, workdir string, env []string) (Runner, error) {
+	return newCmdRunner("perl", code, workdir, env)
+}
+
+// GetOutputScanners is a helper method that returns two bufio.Scanner objects,
+// one for stdout and one for stderr.
+func GetOutputScanners(r Runner) (stdoutScanner bufio.Scanner, stderrScanner bufio.Scanner, err error) {
```

Comment:
I think, you don't need to use `bufio.Scanner` here, as the only usage (so far) for the scanned output is simply print it to the stdout (in `chain.go`). Scanner is a higher level abstraction for tokens of data (and I suppose, `runner`s do not have any assumtions about the data that will be printed), it requires more memory and, well, it's just a new level between the task producing output and the actual printing to stdout.

Moreover, Scanner will fail if the text is bigger that 64K (it's documented, and it's the reasonable choice for most of the scanner uses). So if any task will dump 64K+1 bytes of text, dog will fail.

My suggestion is to use [io.MultiReader](https://golang.org/pkg/io/#MultiReader) to combine both, `stdoutReader` and `stderrReader` into one, and then just use [io.Copy](https://golang.org/pkg/io/#Copy) to print it to the stdout in one goroutine on chain.go. MultiReader will return EOF as soon as all readers return EOF, which is exactly what we want.